kythe/cxx/indexer/cxx/KytheGraphObserver.cc

```diff
@@ -1363,23 +1364,33 @@ const KytheClaimToken* KytheGraphObserver::getAnonymousNamespaceClaimToken(
     CHECK(main_source_file_token_ != nullptr);
     return main_source_file_token_;
   }
-  return getNamespaceClaimToken(loc);
+  return &getNamespaceTokens(loc).anonymous;
 }
 
 const KytheClaimToken* KytheGraphObserver::getNamespaceClaimToken(
     clang::SourceLocation loc) const {
+  return &getNamespaceTokens(loc).named;
+}
+
+const KytheGraphObserver::NamespaceTokens&
+KytheGraphObserver::getNamespaceTokens(clang::SourceLocation loc) const {
   auto* file_token = getClaimTokenForLocation(loc);
-  auto token = namespace_tokens_.find(file_token);
-  if (token != namespace_tokens_.end()) {
-    return &token->second;
-  }
-  proto::VName vname;
-  vname.set_corpus(file_token->vname().corpus());
-  KytheClaimToken new_token;
-  new_token.set_vname(vname);
-  new_token.set_rough_claimed(file_token->rough_claimed());
-  namespace_tokens_.emplace(file_token, new_token);
-  return &namespace_tokens_.find(file_token)->second;
+  auto [iter, inserted] =
+      namespace_tokens_.emplace(file_token, NamespaceTokens{});
+  if (inserted) {
+    // Named namespaces belong to the same corpus as structural types due to
+    // their use as extension points which may be opened from a file in any
+    // corpus, but should still refer to the same node.
+    iter->second.named.mutable_vname()->set_corpus(
+        type_token_.vname().corpus());
+    iter->second.named.set_rough_claimed(file_token->rough_claimed());
+    // Anonymous namespaces are unique to the translation in which they're
+    // defined, which we approximate by using the file.
```

Comment:
"the file's corpus" or "the file_token's corpus" or "the file's vname's corpus".